protocol/contracts/templegold/TempleGoldStaking.sol

```diff
@@ -0,0 +1,521 @@
+ */
+contract TempleGoldStaking is ITempleGoldStaking, TempleElevatedAccess, Pausable {
+    using SafeERC20 for IERC20;
+
+    /// @notice The staking token. Temple
+    IERC20 public immutable override stakingToken;
+    /// @notice Reward token. Temple Gold
+    IERC20 public immutable override rewardToken;
+
+    /// @notice Distribution starter
+    address public override distributionStarter;
+    /// @notice Week length
+    uint256 constant public WEEK_LENGTH = 7 days;
+    /// @notice Total supply of staking token
+    uint256 public override totalSupply;
+
+    /// @notice Time tracking
+    uint256 public override periodFinish;
+    /// @notice Last rewards update time
+    uint256 public override lastUpdateTime;
+
+    /// @notice Store next reward amount for next epoch
+    uint256 public override nextRewardAmount;
+    /// @notice Duration for rewards distribution
+    uint256 public override rewardDuration;
+    /// @notice Cooldown time before next distribution of rewards
+    /// @dev If set to zero, rewards distribution is callable any time 
+    uint160 public override rewardDistributionCoolDown;
+    /// @notice Timestamp for last reward notification
+    uint96 public override lastRewardNotificationTimestamp;
+    /// @notice Cooldown time before unstake
+    uint32 public override unstakeCooldown;
+
+    /// @notice For use when migrating to a new staking contract if TGLD changes.
+    address public override migrator;
+    /// @notice Data struct for rewards
+    Reward internal rewardData;
+    /// @notice Staker balances
+    mapping(address account => uint256 balance) private _balances;
+    /// @notice Account vote delegates
+    mapping(address account => address delegate) public delegates;
+    /// @notice Stakers claimable rewards
+    mapping(address account => uint256 amount) public override claimableRewards;
+    /// @notice Staker reward per token paid
+    mapping(address account => uint256 amount) public override userRewardPerTokenPaid;
+    /// @notice Track voting checkpoints
+    mapping(address account => mapping(uint256 epoch => Checkpoint)) private _checkpoints;
+    /// @notice Track number of checkpoint for account
+    mapping(address account => uint256 number) public override numCheckpoints;
+    /// @notice Track stake times for accounts
+    mapping(address account => uint256 stakeTime) public override stakeTimes;
+    
+    constructor(
+        address _rescuer,
+        address _executor,
+        address _stakingToken,
+        address _rewardToken
+    ) TempleElevatedAccess(_rescuer, _executor){
+        stakingToken = IERC20(_stakingToken);
+        rewardToken = IERC20(_rewardToken);
+    }
+
+    /**
+     * @notice Set unstake cooldown
+     * @param _period Cooldown period. Zero value accepted to disable cooldown feature
+     */
+    function setUnstakeCooldown(uint32 _period) external override onlyElevatedAccess {
+        unstakeCooldown = _period;
```

Comment:
This function has no constraints. ElevatedAccess can block tokens forever and never let users withdraw. 
Consider adding a `maxUnstakeCooldown` period here, so that `_period < maxUnstakeCooldown`
 (to build trust from stakers).